### PR TITLE
[SPARK-16853][SQL] fixes encoder error in DataSet typed select

### PR DESCRIPTION
**Pull request by clockfly**
## What changes were proposed in this pull request?

For DataSet typed select:

```
def select[U1: Encoder](c1: TypedColumn[T, U1]): Dataset[U1]
```

If type T is a case class or a tuple class that is not atomic, the resulting logical plan's schema will mismatch with `Dataset[T]` encoder's schema, which will cause encoder error and throw AnalysisException.
### Before change:

```
scala> case class A(a: Int, b: Int)
scala> Seq((0, A(1,2))).toDS.select($"_2".as[A])
org.apache.spark.sql.AnalysisException: cannot resolve '`a`' given input columns: [_2];
..
```
### After change:

```
scala> case class A(a: Int, b: Int)
scala> Seq((0, A(1,2))).toDS.select($"_2".as[A]).show
+---+---+
|  a|  b|
+---+---+
|  1|  2|
+---+---+
```
## How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by cloud-fan** [comment]
LGTM, though the test haven't passed.

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 1]
```diff
@@ -184,6 +184,17 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
       2, 3, 4)
   }
 
+  test("SPARK-16853: select, case class and tuple") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: how about `typed select that returns case class or tuple`?

Comment (clockfly) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: @cloud-fan 

This follows existing test case name, for example, for select 2, current name is
```
test("select 2, primitive and tuple") {
```

https://github.com/apache/spark/pull/14474/files#diff-3836bd1fc5ae9c8b8ac4bdb2b9944159L196

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by rxin** [comment]
Is this a bug fix? If yes, shouldn't we merge it in 2.0?

**Comment by cloud-fan** [comment]
it's a bug since the very beginning, should we merge it to 1.6 too?

**Comment by zsxwing** [comment]
Should this one be merged into branch-2.0?

**Comment by cloud-fan** [comment]
@clockfly can you create a new PR against 2.0? thanks!

**Pull request closed without merging**